src/drivers/intel/dmic/dmic_nhlt.c

```diff
@@ -0,0 +1,418 @@
+// SPDX-License-Identifier: BSD-3-Clause
+//
+// Copyright(c) 2021 Intel Corporation. All rights reserved.
+//
+// Author: Seppo Ingalsuo <seppo.ingalsuo@linux.intel.com>
+
+#include <sof/drivers/dmic.h>
+#include <sof/math/numbers.h>
+#include <ipc/dai.h>
+#include <ipc/dai-intel.h>
+#include <stdint.h>
+
+#if CONFIG_INTEL_DMIC_NHLT
+
+#include "nhlt_dmic_cfg.h"
+
+/* Base addresses (in PDM scope) of 2ch PDM controllers and coefficient RAM. */
+static const uint32_t base[4] = {PDM0, PDM1, PDM2, PDM3};
+static const uint32_t coef_base_a[4] = {PDM0_COEFFICIENT_A, PDM1_COEFFICIENT_A,
+					PDM2_COEFFICIENT_A, PDM3_COEFFICIENT_A};
+static const uint32_t coef_base_b[4] = {PDM0_COEFFICIENT_B, PDM1_COEFFICIENT_B,
+					PDM2_COEFFICIENT_B, PDM3_COEFFICIENT_B};
+
+static int count_bits_set(uint32_t w)
+{
+	int val = 0;
+	int i;
+
+	for (i = 0; i < 32; i++) {
+		val += w & 1;
+		w = w >> 1;
+	}
+
+	return val;
+}
+
+static int nhlt_dmic_dai_params_get(struct dai *dai, uint32_t outcontrol[],
+				    struct nhlt_pdm_ctrl_cfg *pdm_cfg[],
+				    struct nhlt_pdm_ctrl_fir_cfg *fir_cfg[])
+{
+	struct dmic_pdata *dmic = dai_get_drvdata(dai);
+	int fir_stereo = FIR_CONTROL_A_STEREO_GET(fir_cfg[0]->fir_control);
```

Comment:
It can be useful for changes needed with next platform, I'd prefer to keep it as-is. Now the functions gets all driver HW parts related information that might impact what DAI reports.